### PR TITLE
Agolubev #16967 add input size attribute

### PR DESCRIPTION
**Pull request by agolubev**
added buffer attribute to head and take. Take code is not so beautiful though.

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 1]
```diff
@@ -299,11 +297,13 @@ private[akka] object ActorProcessorFactory {
     // Also, otherwise the attributes will not affect the settings properly!
     val settings = materializer.effectiveSettings(att)
     def interp(s: Stage[_, _]): (Props, Unit) = (ActorInterpreter.props(settings, List(s), materializer, att), ())
+    def interpAttr(s: Stage[_, _], newAttributes: Attributes): (Props, Unit) = (ActorInterpreter.props(settings, List(s), materializer, newAttributes), ())
     op match {
-      case Map(f, _)                  ⇒ interp(fusing.Map(f, settings.supervisionDecider))
-      case Filter(p, _)               ⇒ interp(fusing.Filter(p, settings.supervisionDecider))
-      case Drop(n, _)                 ⇒ interp(fusing.Drop(n))
-      case Take(n, _)                 ⇒ interp(fusing.Take(n))
+      case Map(f, _)    ⇒ interp(fusing.Map(f, settings.supervisionDecider))
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala]: fix formatting

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 2]
```diff
@@ -299,11 +297,13 @@ private[akka] object ActorProcessorFactory {
...
+      case Filter(p, _) ⇒ interp(fusing.Filter(p, settings.supervisionDecider))
+      case Drop(n, _)   ⇒ interp(fusing.Drop(n))
+      case Take(n, _) ⇒ interpAttr(fusing.Take(n),
+        if (n <= materializer.settings.maxInputBufferSize) inputBuffer(initial = n.toInt, max = n.toInt) and att else att)
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala]: n can be non-positive, in that case you still need a buffer of 1.

**Comment by drewhk** [comment]
LGTM after non-positive N is handled by take()

**Comment by ktoso** [comment]
Sadly this needs to be rebased now

**Comment by agolubev** [comment]
Ok, no problem

**Comment by ktoso** [comment]
Thanks in advance :)

**Comment by agolubev** [comment]
My tests are passed, now will see if build will be ok as well

**Comment by drewhk** [comment]
Merged :)

**Pull request merged**